from pypers.steps.qiime import Qiime

class AlphaDiversity(Qiime):
    spec = {
        "version": "2015.05.20",
        "descr": [
            "Calculates the alpha (within-sample) diversity on a directory of OTU tables generated by MultipleRarefactions",
            "Generates output metrics for each sample"
        ],
        "url" : "http://qiime.org/scripts/alpha_diversity.html",
        "args":
        {
            "inputs": [
                {
                    "name"     : "input_dir",
                    "type"     : "dir",
                    "iterable" : True,
                    "descr"    : "the input .biom OTU tables dictionary",
                }
            ],
            "outputs": [
                {
                    "name"  : "output_metrics_dir",
                    "type"  : "dir",
                    "descr" : "output directory of .biom rarerefied (subsampled) OTU tables"
                }
            ],
            "params": [
                {
                    "name"    : "tree",
                    "descr"   : "path to newick tree file, required for phylogenetic metrics",
                    "type"    : "file",
                    "value"   : "/Public_data/microbiome/16s/greengenes/gg_13_8_otus/trees/99_otus.tree",
                    "readonly": True
                }
            ]
        },
        "requirements": { },
    }

    def process(self):

        self.output_metrics_dir = '%s/metrics' % self.output_dir

        cmd = 'alpha_diversity.py -t %s -i %s -o %s' % (self.tree, self.input_dir, self.output_metrics_dir)
        self.submit_cmd(cmd,extra_env=self.extra_env)

