from pypers.core.step import CmdLineStep

class ParallelAlphaDiversity(CmdLineStep):
    spec = {
        "version": "2015.05.20",
        "descr": [
            "Calculates the alpha (within-sample) diversity on a directory of OTU tables generated by ParallelMultipleRarefactions",
            "Generates output metrics for each sample",
            "Warning: We have had problem with this step stalling on the cluster, so prefer to use non-parallel step AlphaDiversity"
        ],
        "url" : "http://qiime.org/scripts/parallel_alpha_diversity.html",
        "args":
        {
            "inputs": [
                {
                    "name"     : "input_dir",
                    "type"     : "dir",
                    "iterable" : True,
                    "descr"    : "the input .biom OTU tables dictionary",
                }
            ],
            "outputs": [
                # The output is the dir not any single file
            ],
            "params": [
                {
                    "name"    : "tree",
                    "descr"   : "path to newick tree file, required for phylogenetic metrics",
                    "type"    : "file",
                    "value"   : "/Public_data/microbiome/16s/greengenes/gg_13_8_otus/trees/99_otus.tree"
                }
            ]
        },
        "requirements": { 
            'cpus' : '10'
        },
        "extra_env" : {
            "PYTHONPATH" : "/software/pypers/qiime/qiime-1.8.0/lib/python2.7/site-packages",
            "PATH": "/software/pypers/qiime/qiime-1.8.0/bin"
        },
        "cmd": [
            "parallel_alpha_diversity.py -T  -t {{tree}} -i {{input_dir}} -o {{output_dir}} --jobs_to_start {{cpus}}"
        ]
    }
